### PR TITLE
refactor command line arguments into subcommands with seperate arguments...

### LINKED ISSUE #285: command line help needs to match command line design
Over in [this issue](https://github.com/letsencrypt/lets-encrypt-preview/issues/136) we have a design for the major verbs and different subcategories of options that the client needs to expose in order to be as swiftly comprehensible as possible to users.

Currently the client, if invoked with --help, prints this:

```
usage: letsencrypt [-h] [-d DOMAIN [DOMAIN ...]] [-s SERVER] [-k AUTHKEY]
                   [-B N] [-R] [--revoke-certificate REV_CERT]
                   [--revoke-key REV_KEY] [-b N] [-v] [-r] [--no-confirm] [-e]
                   [-t] [--config-dir CONFIG_DIR] [--work-dir WORK_DIR]
                   [--backup-dir BACKUP_DIR] [--key-dir KEY_DIR]
                   [--cert-dir CERT_DIR] [--le-vhost-ext LE_VHOST_EXT]
                   [--cert-path CERT_PATH] [--chain-path CHAIN_PATH]
                   [--apache-server-root APACHE_SERVER_ROOT]
                   [--apache-mod-ssl-conf APACHE_MOD_SSL_CONF]
                   [--apache-ctl APACHE_CTL] [--apache-enmod APACHE_ENMOD]
                   [--apache-init-script APACHE_INIT_SCRIPT]
                   [-c FILE [FILE ...]] [--export-conf-file [FILE]]
```
followed by a very long and seemingly randomly-ordered list of options.

It should list the major command verbs at the top, with [options] for each one, not an exhaustive list.  Then enumerate the most important options, and save the subcategories of options for if the user runs with <tt>-h -h</tt>, or asks for help about options for a specific verb.

### PR DESCRIPTION
**Pull request by cooperq**
.... fixes #285

### CODE REVIEW

**Comment by cooperq** [comment]
cc @jdkasten for review

**Comment by jdkasten** [comment]
Thanks @cooperq!  My plan is to try to merge #320 then I will checkout this PR and update it to master. I should make the rounds by tonight.

File: letsencrypt/scripts/main.py [hunk 1]
```diff
@@ -26,75 +26,78 @@
...
-    add("--apache-init-script", default="/etc/init.d/apache2",
+    apply_on_all(CONFIGURATION_PARSERS, "add_argument", "--apache-ctl", default="apache2ctl", help=config_help("apache_ctl"))
+    apply_on_all(CONFIGURATION_PARSERS, "add_argument", "--apache-enmod", default="a2enmod", help=config_help("apache_enmod"))
+    apply_on_all(CONFIGURATION_PARSERS, "add_argument", "--apache-init-script", default="/etc/init.d/apache2",
         help=config_help("apache_init_script"))
+    apply_on_all(CONFIGURATION_PARSERS, "add_argument", "--le-vhost-ext", default="-le-ssl.conf",
+        help=config_help("le_vhost_ext"))
+    apply_on_all(CONFIGURATION_PARSERS, "add_argument", "-b", "--rollback", type=int, default=0, metavar="N",
+        help="Revert configuration N number of checkpoints.")
+    apply_on_all(CONFIGURATION_PARSERS, "add_argument", "-c", "--view-config-changes", action="store_true",
```
Comment (kuba) [inline letsencrypt/scripts/main.py]: This conflics with ConfArgParse's `-c/--config`:
```
$ letsencrypt
usage: letsencrypt [-h] [-d DOMAIN [DOMAIN ...]] [-s SERVER] [-t]
                   {apache, standalone, revoke} ...
letsencrypt: error: argument -c/--view-config-changes: conflicting option string(s): -c
```

File: letsencrypt/scripts/main.py [hunk 2]
```diff
@@ -122,27 +132,45 @@ def main():  # pylint: disable=too-many-branches, too-many-statements
...
+            client.revoke(config, args.no_confirm, args.rev_cert, args.rev_key)
+            sys.exit()
+    except AttributeError: 
+        pass
 
-    if args.revoke or args.rev_cert is not None or args.rev_key is not None:
-        client.revoke(config, args.no_confirm, args.rev_cert, args.rev_key)
-        sys.exit()
+    try:
+        if args.rollback > 0:
```
Comment (kuba) [inline letsencrypt/scripts/main.py]: how about having a `lestsencrypt rollback` command?

**Comment by kuba** [comment]
Since all Travis tests fail (functional and lint/style), and the binary fails in practice as well (https://github.com/letsencrypt/lets-encrypt-preview/pull/322#discussion_r27278374), I assume this is an very early iteration and a request for architectural feedback.

1. `letsencrypt apache/standalone` syntax doesn't really scale under the new plugin architecture from #320. How about `letsencrypt run --auth apache/standalone ...`.? We could later extend it to `letsencrypt run --auth auth_entry_point_name --installer installer_entry_point_name`
2. `IConfig` was meant to represent user-configurable state and it now mostly proxies to `constants`. Instead of doing that, all calls to `config.foo` should be changed to `constants.FOO`.
3. Despite using constants (e.g. `APACHE_CTL`) you kept corresponding CLI arguments (`--apache-ctl`). That's extremely confusing, since those CLI argument are de facto ignored!
4. Anyway, why do you want to move all of those to `constants.py`? We're making a full circle back to state before #220, making most of the stuff not configurable at all... Is it merely because noisy output in `letsencrypt --help`? If so, we could easily suppress most of the output...

**Comment by kuba** [comment]
Also, re "full circle", please note, that ConfArgParse is actually allowing users to have a config file at an arbitrary location. We shouldn't expect users to edit Python file that's part of our package. Consider a case where, letsencrypt is packaged for Debian (in contrast to working with git repo copy): users should not modify any system files from /usr/local, where the package would be installed.

File: letsencrypt/scripts/main.py (hunk 1 above)
Comment (cooperq) [inline letsencrypt/scripts/main.py]: I vote for removing the short argument entirely in this case. The previous short argument was -v which many users will confuse for `verbose` or possibly `version`

**Comment by cooperq** [comment]
It is annoying that the travis tests fail since they don't fail on my local machine. I will need to look into why that is happening. As for your other points I will address them one by one. 
1. It might make more sense to have the following subcommands for LE `generate, configure, revoke, rollback` and specify the target as an option. I think the default option would be `configure` which would of course generate and configure a certificate. 
2. In that case I think that we should move many of the things in `constants` to `IConfig` instead. This is simply a case of me not being really sure where to put these. 
3. See below
4. The reason for moving all of these into constants was to make global defaults. The problem of using subparsers in argparse, like I am doing, is that only the subparser and it's parent options get put into the resulting args object. Currently LE is architected in such a way that it expects all of the args to be there and have defaults.  Perhaps a better workaround is to set defaults arguments in `IConfig` instead of `constants`

**Comment by kuba** [comment]
I have quickly drafted a version with plugin discovery and following subcommands:

```
run                 Obtain a certificate and install.
auth                Obtain a certificate (no install).
install             Install (no auth).
revoke              Revoke.
rollback            Rollback.
config_changes      View config changes.
plugins             List plugins.
```

Branch available at https://github.com/kuba/lets-encrypt-preview/tree/cli, and you can view changes at
https://github.com/letsencrypt/lets-encrypt-preview/compare/master...kuba:cli

**Comment by jdkasten** [comment]
Another reason for moving work_dir config_dir ect... away from the command-line, is because we are going to experience **major difficulties** if anyone changes it other than the distributors.  Especially as we work through problems with renewal... these variables really need to be set and be the same for every single run of the client.  Plus, no one is going to understand the consequences of changing those variables and mistakes will not be easy to fix.  (I don't want to spend the time on forums in the future trying to help people repair their stuff :) )

**Comment by kuba** [comment]
I have just synced https://github.com/kuba/lets-encrypt-preview/tree/cli with all my local developments. Unfortunately, I'm really short on time at the moment, so anyone is welcome to take off from there :)

**Comment by kuba** [comment]
Superseded by #376.

**Pull request closed without merging**